drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,757 @@
+		return -ENOMEM;
+	}
+#endif
+
+	if (direction == SDIO_IO_WRITE) {
+		if (data->data) {
+#ifdef CONFIG_SDHC_STM32_POLLING_SUPPORT
+			res = HAL_SDIO_WriteExtended(config->hsd, &arg, data->data,
+						     dev_data->total_transfer_bytes,
+						     data->timeout_ms);
+#else
+			memcpy(dev_data->sdio_dma_buf, data->data, dev_data->total_transfer_bytes);
+			sys_cache_data_flush_range((uint32_t *)dev_data->sdio_dma_buf,
+						dev_data->total_transfer_bytes);
+			res = HAL_SDIO_WriteExtended_DMA(config->hsd, &arg, dev_data->sdio_dma_buf,
+							 dev_data->total_transfer_bytes);
+#endif
+		} else {
+			LOG_ERR("Invalid data buffer passed to CMD53 (write).");
+			return -EINVAL;
+		}
+	} else {
+#ifdef CONFIG_SDHC_STM32_POLLING_SUPPORT
+		res = HAL_SDIO_ReadExtended(config->hsd, &arg, data->data,
+					    dev_data->total_transfer_bytes, data->timeout_ms);
+#else
+		sys_cache_data_flush_range((uint32_t *)dev_data->sdio_dma_buf,
+					dev_data->total_transfer_bytes);
+		res = HAL_SDIO_ReadExtended_DMA(config->hsd, &arg, dev_data->sdio_dma_buf,
+						dev_data->total_transfer_bytes);
+#endif
+	}
+
+#ifndef CONFIG_SDHC_STM32_POLLING_SUPPORT
+	/* Wait for whole transfer to complete */
+	if (k_sem_take(&dev_data->device_sync_sem, K_MSEC(CONFIG_SD_CMD_TIMEOUT)) != 0) {
+		return -ETIMEDOUT;
+	}
+	if (direction == SDIO_IO_READ) {
+		memcpy(data->data, dev_data->sdio_dma_buf, data->block_size * data->blocks);
+	}
+	k_free(dev_data->sdio_dma_buf);
+#endif
+
+	return res == HAL_OK ? 0 : -EIO;
+}
+
+static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+{
+	uint32_t res = 0;
+	struct sdhc_stm32_data *data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	/* Check if host supports 1.8V signaling */
+	if (!data->props.host_caps.vol_180_support) {
+		LOG_ERR("Host does not support 1.8V signaling");
+		return -ENOTSUP;
+	}
+
+	res = SDMMC_CmdVoltageSwitch(config->hsd->Instance);
+	if (res != 0) {
+		LOG_ERR("CMD11 failed: %u", res);
+		return -EIO;
+	}
+
+	LOG_INF("Successfully switched to 1.8V signaling");
+	return 0;
+}
+
+static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd,
+			      struct sdhc_data *data)
+{
+	int res;
+	struct sdhc_stm32_data *dev_data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	__ASSERT(cmd != NULL, "Command is NULL.");
+
+	if (!WAIT_FOR(HAL_SDIO_GetState(config->hsd) == HAL_SDIO_STATE_READY, SD_TIMEOUT_US,
+		      k_usleep(1))) {
+		LOG_ERR("SD Card is busy now");
+		return -EBUSY;
+	}
+
+	(void)pm_device_runtime_get(dev);
+
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&dev_data->bus_mutex, K_FOREVER);
+
+	dev_data->cmd_index = cmd->opcode;
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+		res = sdhc_stm32_go_idle_state(dev);
+		break;
+
+	case SD_SELECT_CARD:
+		res = SDMMC_CmdSelDesel(config->hsd->Instance, cmd->arg);
```

Comment:
`SDMMC_CmdSelDesel()` return value is not a standard *errno* whereas this function is expected to return one.
Ditto for lines 395, 402 below.